apps/app.go

```diff
@@ -73,3 +94,77 @@ func ParseApp(jsonBlob []byte) (*App, error) {
 
 	return &wrapper.App, err
 }
+
+type RegistrationIntent struct {
+	Name      string
+	PortIndex int
+	Tags      []string
+}
+
+func (app *App) RegistrationIntents(nameSeparator string) []*RegistrationIntent {
+	commonTags := labelsToTags(app.Labels)
+
+	definitions := app.findConsulPortDefinitions()
+	if len(definitions) == 0 {
+		return []*RegistrationIntent{
+			&RegistrationIntent{
+				Name:      app.labelsToName(app.Labels, nameSeparator),
+				PortIndex: 0,
+				Tags:      commonTags,
+			},
+		}
+	}
+
+	var intents []*RegistrationIntent
+	for _, d := range definitions {
+		intents = append(intents, &RegistrationIntent{
+			Name:      app.labelsToName(d.Labels, nameSeparator),
+			PortIndex: d.PortIndex,
+			Tags:      append(commonTags, labelsToTags(d.Labels)...),
+		})
+	}
+	return intents
+}
+
+func marathonAppNameToServiceName(name string, nameSeparator string) string {
+	return strings.Replace(strings.Trim(strings.TrimSpace(name), "/"), "/", nameSeparator, -1)
+}
+
+func labelsToTags(labels map[string]string) []string {
+	tags := []string{}
+	for key, value := range labels {
+		if value == "tag" {
+			tags = append(tags, key)
+		}
+	}
+	return tags
+}
+
+func (app *App) labelsToName(labels map[string]string, nameSeparator string) string {
+	appConsulName := app.labelsToRawName(labels)
+	serviceName := marathonAppNameToServiceName(appConsulName, nameSeparator)
+	if serviceName == "" {
+		log.WithField("AppId", app.ID.String()).WithField("ConsulServiceName", appConsulName).
+		Warn("Warning! Invalid Consul service name provided for app. Will use default app name instead.")
+		return marathonAppNameToServiceName(app.ID.String(), nameSeparator)
+	}
+	return serviceName
+}
+
+type IndexedPortDefinition struct {
+	PortIndex int
+	Labels    map[string]string
+}
+
+func (app *App) findConsulPortDefinitions() []IndexedPortDefinition {
+	var definitions []IndexedPortDefinition
+	for i, d := range app.PortDefinitions {
+		if _, ok := d.Labels[MARATHON_CONSUL_LABEL]; ok {
+			definitions = append(definitions, IndexedPortDefinition{
+				PortIndex: i,
```

Comment:
In `portDefinitions` you can either provide a port explicitly or provide zero in which case the port will be randomly picked. I've added support for the latter only. I guess you're right that we should support both ways.